docs/references/architecture/adr-113-mempool-lanes.md

```diff
@@ -0,0 +1,312 @@
+* Timebox this: 1-2 days each
+
+
+## Decision
+
+TODO
+
+## Detailed Design
+
+TO DECIDE:
+
+* bytes per tx, \# of txs, total # of bytes: shared?
+  * Current view: **can be different** easily
+  * MVP is viable without it, but use cases less customizable
+  * Update: How about Injective's use case?
+
+* Two options for introducing several TX lists:
+  1. Multiple `Clist`s inside `CListMempool`, the cache stays inside `CListMempool`.
+  2. Multiple `CListMempool` instances, the cache needs to be outside `CListMempool` (likely, directly controlled by the reactor)
+
+* After discussion with Osmosis and Injective, there's a dilemma.
+  * For Osmosis, a two-lane solution ("native" and "priority") would probably be enough.
+    * This means that we could defer the lane and priority configuration, and not tackle it as part of the MVP
+  * For Injective, their use case seems to be more demanding: many lanes (4?, 9?),
+    with one lane even having less prio than the native lane
+    * If we want to address that use case fully, our MVP _does need_ to tackle the design of lane & prio configuration
+    * so, a more ambitious MVP, implying more energy, time, and resources. We must be careful with what we promise for when
+
+* Byzantine nodes may gossip transactions on a high-priority lane to get preferential treatment.
+  * we may consider banning peers that send transactions with incorrect lane information.
+  * Decision was not to ship lane info, but we're still using different channels for different lanes
+    * **What shall we do when we see a peer misusing a p2p channel?**
+      * not clear what to do, will heavily depend on use case. So
+        * let's wait for someone to hit this so we have a use case to solve
+        * and less things to do for MVP
+      * Updated discussion, looks like the way to go is:
+        * Do "stop peer for error" in any case
+        * Defer decision about banning policy until a use case pops up
+
+
+### Decisions on Minimum Viable Product (MVP)
+
+A lane can be viewed as a mempool on its own, with its own config, capacity, cache,
+gossip protocol, p2p channel, p2p bandwidth capacity, etc.
+For an MVP, probably some (most?) of these components can be common to all lanes.
+In further versions, we can see which of these need to be per-lane.
+What is clear is that each lane will need its own `CList` data structure, right from the beginning.
+* `CList` **need to be different**
+  * if using a common `CList`, the broadcast routine becomes overly complicated,
+    whereas each lane having its own `CList`, the broadcast routine is very simple
+    (advance the highest priority, non-empty lane)
+* cache **can be shared** in MVP
+  * lock contention: we think it won't be a big deal performance-wise for an MVP
+  * cache eviction of entries: lanes with a lot of traffic may force eviction on TXs
+    on lanes with little traffic.
+    * For an SDK application this is irrelevant: SDK implements replay protection (nonce)
```

Comment:
Is it relevant for non-SDK? In general we should also think about non-SDK, ideally we would have a non-SDK "customer" or reviewer.

Reply:
It's irrelevant to any _serious_ chain, as the problem of replay protection MUST be solved at app level, so serious apps should have a mechanism. The cache in the mempool is just a partial solution that helps with performance but that doesn't offer absolute guarantees.

> ideally we would have a non-SDK "customer" or reviewer.

Totally agree, although TTBOMK no non-SDK apps have come up so far with TX dissemination problems that would need QoS